### DISCUSSION TITLE
SocialClub ROS Message

### DISCUSSION THREAD

**Discussion by Yimura** [YimMenu/YimMenu#386] [Research]
## The signature

You need to search for the following signature:
```cpp
E8 ? ? ? ? EB 15 48 8D 55 97 // .add(1).rip()
```

This function is not in the `GTA5.exe` module, you need to scan for this address inside of `socialclub.dll`.

## The function (hook)
I forgot to make a screenshot of the function so I can only show my own implementation of the hook.

The purpose of the original function is to check if the PresenData object is the one we want to parse, it takes the following arguments:
 - **a1**: Presence Object
 - **a2**: Presence Name
 - **a3**: Bool always true according to my limit xreffing this function

The function returns true or false, **TRUE** if the Presence Object matches the Presence Name passed.
The function in question does not modify the presence object so calling it isn't harmful to the original object.

### Class Used
```cpp
#pragma pack(push, 4)
class PresenceData
{
public:
    std::uint64_t m_base_addr;
    char gap8[4];
    int m_offset;
    int m_prefix_len;

    // inline helper I made based on the code in the function we're hooking
    inline const char* event_data() const
    {
        return reinterpret_cast<const char*>(m_base_addr + m_offset);
    }
};
#pragma pack(pop)
```
The return value of the `PresenceData::event_data()` returns a string that looks something like this:
```
ros.friends.invite.received":{"gh":"SC <removed RID>","nickname":"<removed Nick>","avatarurl":""}
```
That's why in the below code you'll see me do some pointer math to get the proper start of the JSON string.

### Possible Events
This function probably sees the following ROS Messages, I did not check if some of these weren't used but it should give you an idea of what might come through here.
![image](https://user-images.githubusercontent.com/24669514/182644200-3982b538-c017-4f50-9951-af622fa81b2c.png)

### Hook
```cpp
bool hooks::is_presence_actual(void* a1, const char* presence_name, bool always_true)
{
    const auto result = g_hooking->m_is_presence_actual.get_original<decltype(&is_presence_actual)>()(a1, presence_name, always_true);
    // only do our stuff once the original returns true
    if (!result)
        return result;

    const auto presence = static_cast<PresenceData*>(a1);
    const auto data = presence->event_data();

    auto body = nlohmann::json::parse(data + presence->m_prefix_len + 2);
    switch (rage::joaat(presence_name))
    {
    case RAGE_JOAAT("ros.publish"):
    {
        break;
    }
    default:
    {
        LOG(INFO) << "Presence[" << presence_name << "]: " << body.dump(4);
        break;
    }
    }

    return result;
}
```

## What use do we have by hooking this?
Returning false will suppress the code in socialclub.dll from handling the ROS Message.

You can really only use the hook for information. Since the invite spam has been patched I don't really see a reason to block ROS Messages.

**Comment by Vali0004** [comment]
I mean, we could use it for notifies when a player tries to join. It'd also be good at blocking game invites and other notifications people want to say "fuck off" to.

**Reply by Yimura** [reply]
> I mean, we could use it for notifies when a player tries to join. It'd also be good at blocking game invites and other notifications people want to say "fuck off" to.

Thing is it doesn't notify on players joining or requesting to join your session, you can only use it to suppress notifications for crew invites, MP invites and others. R* completely made this useless after someone abused it.

**Reply by Vali0004** [reply]
> > I mean, we could use it for notifies when a player tries to join. It'd also be good at blocking game invites and other notifications people want to say "fuck off" to.
> 
> Thing is it doesn't notify on players joining or requesting to join your session, you can only use it to suppress notifications for crew invites, MP invites and others. R* completely made this useless after someone abused it.

Yeah, but those still drive me fucking nuts.

**Reply by hackerftsg** [reply]
> > > I mean, we could use it for notifies when a player tries to join. It'd also be good at blocking game invites and other notifications people want to say "fuck off" to.
> > 
> > 
> > Thing is it doesn't notify on players joining or requesting to join your session, you can only use it to suppress notifications for crew invites, MP invites and others. R* completely made this useless after someone abused it.
> 
> Yeah, but those still drive me fucking nuts.

If you want to replicate it, I would suggest you take a look at this code which has the ROS encryption reversed that R* uses to send requests through the socialclub.dll module: https://github.com/citizenfx/fivem/blob/master/code/components/ros-patches-five/src/SCUIStub.cpp
